plugin/parser/tokenizer.go

```diff
@@ -733,14 +708,21 @@ func (t *tokenizer) scanTo(delimiter []rune) error {
 }
 
 func (t *tokenizer) scan() {
-	if t.scanner != nil {
-		if t.scanner.Scan() {
-			t.buffer = append(t.buffer, []rune(t.scanner.Text())...)
+	if t.reader != nil {
+		s, err := t.reader.ReadString('\n')
+		readSuccessfully := (err == nil || err == io.EOF)
+		needEOF := (err != nil)
+		returnError := (err != nil && err != io.EOF)
+		if readSuccessfully {
+			t.buffer = append(t.buffer, []rune(s)...)
 			t.len = uint(len(t.buffer))
-		} else {
+		}
+		if needEOF {
+			t.reader = nil
 			t.buffer = append(t.buffer, eofRune)
-			t.scanErr = t.scanner.Err()
-			t.scanner = nil
+		}
+		if returnError {
+			t.readErr = err
```

Comment:
This is a bit hard to catch.

Can we use `if ... else if ... else ...` to make things more clear?

Reply:
The `if ... else if ... else ...` is not suitable here because the three cases overlap.

Reply:
Can we merge the last two if branches?

Reply:
SGTM, how about now?

Reply:
LGTM.

Reply:
I refactor it as `if ... else if ... else ...` 0v0